### PR TITLE
[video] Video Versions: Refactor play/select logic

### PR DESCRIPTION
**Pull request by ksooo**
This is a follow-up to the original video version feature PR:

* Consistency: Move video version play/select action handling into `CVideo(Play|Select)ActionProcessorBase` so that we now should have consistent behaviour across all places where videos (with versions) are selected or played.
* Separation of concerns: Move the play logic outside the video version dialog (accept for its own play button) so that the dialog now has a manage mode (like before) and a select mode (instead of play mode). The caller of the dialog now can control what to do with the selected version. It is no longer hard-wired to play what was selected.
* Consistency: before this PR select dialog only popped up when "Play" was selected from context menu or if select action in the video window / Home Screen widget was "play" - all other actions are executed for the current default video version. This is inconstant, so I changed the behaviour (last commit) to always ask whenever multiple version are available (unless setting "play default version" is activated). If people think this is no behaviour we want to have, we can easily remove the last commit from this PR.
* Consistency: Rename setting 'Play default video version' to 'Select default video version' and move it from Player->Videos to Media->Videos.
* Usability: Add context menu item 'Choose version' for videos with multiple versions.

Note: I was not able to get rid of the ugly hack that to make playback work video database must be patched temporarily, because this is tied so hard with the current implementation, that at least I was not able to fix this with reasonable effort. This was part of the original PR and I wanted to fix this and failed. :-( But I have marked those two places in the code with a `todo` so this can be found and picked up at any time. This should be fixed with another PR.

Runtime-tested on macOS latest Kodi master.

@enen92 as you are quite familiar with the Action processor concept, could you please review.

### CODE REVIEW

File: xbmc/video/dialogs/GUIDialogVideoVersion.cpp
Comment (enen92) [inline xbmc/video/dialogs/GUIDialogVideoVersion.cpp:96]: is this needed? m_cancelled should default to false I guess

File: xbmc/video/guilib/VideoActionProcessorHelper.cpp
Comment (enen92) [inline xbmc/video/guilib/VideoActionProcessorHelper.cpp:35]: typo tempoarily

File: xbmc/video/guilib/VideoActionProcessorHelper.cpp
Comment (enen92) [inline xbmc/video/guilib/VideoActionProcessorHelper.cpp:67]: typo too

File: xbmc/video/guilib/VideoActionProcessorHelper.cpp
Comment (enen92) [inline xbmc/video/guilib/VideoActionProcessorHelper.cpp:22]: Is this 100% video version related? If so does it justify a rename?
Just asking since `CVideoActionProcessorHelper` is a really generic name which will likely serve to dump anything else in the future :)

**Review by enen92 — APPROVED** [review]
I am ok with this. Most of the code is an improvement.
I can't really help much with the db hack -  it's something I lack the context to think about a possible solution

File: xbmc/video/guilib/VideoActionProcessorHelper.cpp
Comment (ksooo) [reply xbmc/video/guilib/VideoActionProcessorHelper.cpp:35]: Will change.

File: xbmc/video/guilib/VideoActionProcessorHelper.cpp
Comment (ksooo) [reply xbmc/video/guilib/VideoActionProcessorHelper.cpp:22]: 100% video

Comment (ksooo) [reply xbmc/video/guilib/VideoActionProcessorHelper.cpp:22]: In a future PR I will move resume selection dialog code which is currently duplicated to this helper, so 100% video but not 100% video version.

File: xbmc/video/dialogs/GUIDialogVideoVersion.cpp
Comment (ksooo) [reply xbmc/video/dialogs/GUIDialogVideoVersion.cpp:96]: The dialog instance survives close and reopen of the dialog so the member needs to be re-inited. Do you have a mother piece in mind to do this?

Comment (enen92) [reply xbmc/video/dialogs/GUIDialogVideoVersion.cpp:96]: Ah right, disregard then

**Comment by ksooo** [comment]
@xodidox for your information. Maybe you want to do a runtime test before we merge this.

**Comment by xodidox** [comment]
> @xodidox for your information. Maybe you want to do a runtime test before we merge this.

Thanks for the change. I'll test it.

**Comment by ksooo** [comment]
jenkins build this please

**Comment by xodidox** [comment]
@ksooo it's playing the wrong item in this condition:

1. Turned on the "show as folder" option for video version
2. You have a resume point for the default version of a movie (e.g., played it for several minutes)
3. Choose the "Play from beginning" context menu item for the movie in media window

The video version selection dialog is expected in this scenario, but it's not and a wrong video is playing.

**Comment by ksooo** [comment]
> Choose the "Play from beginning" context menu item for the movie in media window
> The video version selection dialog is expected in this scenario, but it's not 

Well, not after this PR. All (!) items in the context menu are consistent (!) now, always working with the default video version.  

> and a wrong video is playing.

 Confirming this issue. Default video should be played from beginning, but another version is picked up and played.

**Comment by xodidox** [comment]
> > Choose the "Play from beginning" context menu item for the movie in media window
> > The video version selection dialog is expected in this scenario, but it's not
> 
> Well, not after this PR. All (!) items in the context menu are consistent (!) now, always working with the default video version.

It's due to this change: https://github.com/xbmc/xbmc/pull/24180/files#diff-0c47520befb586aae890973984b0ed4038323f60ed43d7b39c364a19cbd5f2b3L200

**Comment by ksooo** [comment]
> It's due to this change

As I said, not to display the version select dialog is by intention, to get a consistent behaviour.

Picking the wrong video needs to be fixed. I'll have a look.

**Comment by xodidox** [comment]
> > It's due to this change
> 
> As I said, not to display the version select dialog is by intention, to get a consistent behaviour of the context menu.

I see the point. I'm not sure if it's the expected behavior though as there is an option "Play default video version" to control this.

**Comment by ksooo** [comment]
> I see the point. I'm not sure if it's the expected behavior though as there is an option "Play default video version" to control this.

 I think most important is consistency, so that the user understands what the context menu stands for. Current behaviour (before this PR) acting upon the default video version with one exception "Play from beginning" that asks for the version is inconsistent. Either it asks for the version on every context menu item or never.
 
 I understand the "Play default video version" setting (which by the way should be renamed to something like "Automatically select default version") that surpasses the version selector for all actions, not just "play from the beginning".

**Comment by xodidox** [comment]
@ksooo I agree, this seems to be better than current behavior.

**Comment by ksooo** [comment]
@xodidox the issue with selecting the wrong video version is fixed with latest force-push. Thanks for reporting.

**Comment by da-anda** [comment]
> I think most important is consistency, so that the user understands what the context menu stands for. Current behaviour (before this PR) acting upon the default video version with one exception "Play from beginning" that asks for the version is inconsistent. Either it asks for the version on every context menu item or never.
> 
> I understand the "Play default video version" setting (which by the way should be renamed to something like "Automatically select default version") that suppresses the version selector for all actions, not just "play from beginning".

the only exception to NOT use the default version on these actions is when there is a resume point, as the resume action has to resume the version the resume point belongs to and not the default version.

**Comment by ksooo** [comment]
> the only exception to NOT use the default version on these actions is when there is a resume point, as the resume action has to resume the version the resume point belongs to and not the default version.

@da-anda I disagree. There should be no exception, not even for resume.

> when there is a resume point, as the resume action has to resume the version the resume point belongs to and not the default version.

 I agree, and as the context menu always displays the resume point of the default version, there should be no exception.
 
 BTW: I'm going to add a new context menu item "Choose video version" so any version can be picked up easily from the context menu.

File: xbmc/video/guilib/VideoSelectActionProcessor.cpp
Comment (da-anda) [inline xbmc/video/guilib/VideoSelectActionProcessor.cpp:46]: I suppose this is used to prevent endless recursion for the select dialog? As there is already the `choose_video_version` property, wouldn't it be sufficient to simply set this property to `false` once the dialog was shown? Just curious, no request to change anything.

**Comment by da-anda** [comment]
> > when there is a resume point, as the resume action has to resume the version the resume point belongs to and not the default version.
> 
> I agree, and as the context menu always displays the resume point of the default version (if any), there should be no exception.

if the resume point is only shown for the default version in lists, then all is fine and no exception required 👍. I only hope that the "continue watching" home screen widget will show the actual version to resume and not the default version - haven't tested this yet.

> BTW: I'm going to add a new context menu item "Choose video version" so any version can be picked up easily from the context menu.

awesome. I'd probably only call it `Choose version` though

File: xbmc/video/guilib/VideoSelectActionProcessor.cpp
Comment (ksooo) [reply xbmc/video/guilib/VideoSelectActionProcessor.cpp:46]: No. Again, separation of concerns! It's a property of the item. What we want to remember is a property of the processor. Unrelated things.

**Comment by da-anda** [comment]
> Hold yer horses. Nobody said that we are bug free and production ready. I have not checked, but I assume that the widget will always display the default video version's data.

it's partially working, but if you f.e. have resume points on several versions, only one item will be shown in the "continue watching" widget.

> This is btw exactly what I mean with we got the basic design of this feature wrong. A version of a video should imo be modelled as a normal movie, with an additional identifier that is used to combine arbitrary movie items, like we do for movie sets. This approach would have led to much less special cases and pitfalls. For Continue watching it would have worked fully automatically, then, for example.

I see, and I fear that I have to agree with your assessment in Slack now. Hadn't realized that the basic design causes that many issues and still requires a ton of workarounds all over the place to be ready for prime time 😞

**Comment by jjd-uk** [comment]
@ksooo Given the test build a try and all looks good from an Estuary perspective. Also everything I've tried from a use perspective seems to be still ok too.

**Comment by ksooo** [comment]
> I see you've just added the "Choose version" context menu item, so if want me to add an Approval I'll wait to test that first.

@jjd-uk your approval is highly appreciated. I will kick off a test build...

**Comment by jjd-uk** [comment]
@ksooo can you confirm intended behaviour of `Choose version` if a default version is set and `Select default video version` is enabled? In this circumstance the context menu refreshes itself and `Play` context menu item is highlighted, seems like the `Choose version` is redundnant in this case.

**Comment by ksooo** [comment]
The last force-push was just cosmetics. Thanks for reporting back. I will have a look later tonight.

**Comment by ksooo** [comment]
@jjd-uk I fixed the wrong behaviour of `Choose version` if a default version is set and `Select default video version` is enabled. Will kick off another test build and let you know once it is ready.

**Comment by ksooo** [comment]
Jenkins build this please

**Review by garbear — APPROVED** [review]
My feedback:

* Consistency - good refactor of the original code
* Separation of concerns - Another good refactor
* Consistency (pt 2) - I haven't real-time tested this, but it seems low risk, and can be removed if user feedback indicates it's undesirable
* Consistency (pt 3) - The changes to the language strings seems ok
* Usability - The added context menu item is probably beneficial

Regardless, I think as we hammer out these minor improvements, we'll evolve the code to a more acceptable standard, which will inform futher improvements as we work this into a shippable state.

**Comment by ksooo** [comment]
closed/reopened to hopefully get Jenkins build errors resolved...

**Comment by da-anda** [comment]
thanks for all the time you put into improving this feature @ksooo even though you don't really have time for this atm - highly appreciated.

**Comment by xodidox** [comment]
@ksooo I still got the wrong item played when using the "Playing from beginning" context menu when show video version as folder. The issue is:

https://github.com/xbmc/xbmc/blob/master/xbmc/video/VideoUtils.cpp#L414

All items in the video version folder of the movie are added to the playlist, thus the first item instead of the default version is playing.

At the time of above code is called, the item's m_bIsFolder is still true. The logic to use a temporary item and set m_bIsFolder as false is not applied.

**Comment by ksooo** [comment]
@xodidox I guess I will not do another round of digging and merge this as it is now. This PR fixes a lot of things already and we can live with one regression that can be tackled in a follow up PR, esp. as I’m not able to reproduce with the information I currently have.

**Comment by ksooo** [comment]
@xodidox are you sure you tested with the really latest version of this PR? I force-pushed a lot during the last days.

**Comment by ksooo** [comment]
I’m at the phone atm so cannot debug. Will have a look when I’m back home. 

Your hack with patching the folder flag should be still there and applied in videoselectactionhelper::choosevideoversion. Set a breakpoint there and check why the hack is not applied.

**Review by jjd-uk — APPROVED** [review]
The issue I raised is fixed.

Thanks @ksooo

**Comment by ksooo** [comment]
I think I see the problem already. I moved that logic from your dialog to the new helper, but it seems it is buggy. See folder hack gets only applied if auto select default version setting is on and show as folder setting is on. But it must of also be applied if auto select default version is off I guess.

**Comment by ksooo** [comment]
So, it should currently work if auto select default version is on and fail if the setting is off. Can you confirm that?

**Comment by jurialmunkey** [comment]
This is much improved imho. 

My only issue is in the management dialog mode the buttons on the right should be for the focused item and not require the user to first click the item in the list to update. (but that's a pre-existing issue and perhaps for another PR).

Only re-raising the focus vs. click issue now because with the play mode dialog being more prominently used, it's a little more urgent to fix. Can't use the hack of `<onright>Action(Select)</onright>` anymore since that will cause playback in play mode -- and without that ugly hack, the management dialog is currently utterly confusing as to what item the buttons are acting upon.

**Comment by ksooo** [comment]
Should be ready to go now. This is definitely not a master piece of software engineering, but imo definitely an improvement, code- and function-wise. Will merge once Jenkins is green. I tested carefully and found everything working now, but cannot
 guarantee that I have not overlooked something. Some much special handling required, because of the very "special" design of this feature. :-( Let's see how it goes...

**Pull request merged**
